### PR TITLE
Make dotty compile backend.

### PR DESCRIPTION
**Pull request by DarkDimius**
Changes that make dotty one step closer to bootstrap. 
And a fix to Pattern matcher revealed by backend.

### CODE REVIEW

**Comment by DarkDimius** [comment]
Currently fails at erasure with 
```
./src/dotty/tools/backend/jvm/GenBCode.scala:215: warning: cannot merge class PrefixSetting in class MutableSettings with type bounds <: ScalaSettings.super.Setting{T = scala.collection.immutable.List[String]} as members of one type; keeping only class PrefixSetting in class MutableSettings
      lazy val localOpt = new LocalOpt(new Settings())
                              ^
findMember exception for TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing), AndType(RefinedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,settings)),ScalaSettings)),EnableSettings), scala$tools$nsc$settings$MutableSettings$EnableSettings$$T, TypeAlias(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,settings)),ScalaSettings)),BooleanSetting)) | -1632041102),TypeRef(ThisType(TypeRef(NoPrefix,scala)),Singleton))) member scala$tools$nsc$settings$MutableSettings$EnableSettings$$T
findMember exception for TypeRef(NoPrefix,_1103.type) member scala$tools$nsc$settings$MutableSettings$EnableSettings$$T
findMember exception for TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing), AndType(RefinedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,settings)),ScalaSettings)),EnableSettings), scala$tools$nsc$settings$MutableSettings$EnableSettings$$T, TypeAlias(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,settings)),ScalaSettings)),BooleanSetting)) | -1632041102),TypeRef(ThisType(TypeRef(NoPrefix,scala)),Singleton))) member scala$tools$nsc$settings$MutableSettings$EnableSettings$$T
findMember exception for TypeRef(NoPrefix,_1103.type) member scala$tools$nsc$settings$MutableSettings$EnableSettings$$T
exception while typing new tools.nsc.backend.jvm.opt.LocalOpt(new tools.nsc.Settings()) of class class dotty.tools.dotc.ast.Trees$Apply # 1018023
exception while typing lazy <accessor> def localOpt: scala.tools.nsc.backend.jvm.opt.LocalOpt = new tools.nsc.backend.jvm.opt.LocalOpt(new tools.nsc.Settings()) of class class dotty.tools.dotc.ast.Trees$DefDef # 1315805
exception while typing class Worker2() extends Object() {
...
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:600)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:383)
[error]     at dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:1247)
[error]     at dotty.tools.dotc.core.Types$NamedType.denotAt(Types.scala:1224)
[error]     at dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:1212)
[error]     at dotty.tools.dotc.core.Types$NamedType.info(Types.scala:1384)
[error]     at dotty.tools.dotc.core.TypeApplications$.recur$2(TypeApplications.scala:395)
[error]     at dotty.tools.dotc.core.TypeApplications$.recur$2(TypeApplications.scala:402)
[error]     at dotty.tools.dotc.core.TypeApplications$.containsSkolemType$extension(TypeApplications.scala:410)
[error]     at dotty.tools.dotc.core.Types$RefinedType.refinementRefersToThis(Types.scala:1770)
[error]     at dotty.tools.dotc.core.Types$Type.goRefined$1(Types.scala:456)
[error]     at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:425)
[error]     at dotty.tools.dotc.core.Types$Type.goAnd$1(Types.scala:489)
[error]     at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:443)
[error]     at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:491)
[error]     at dotty.tools.dotc.core.Denotations$Denotation.findMember(Denotations.scala:166)
[error]     at dotty.tools.dotc.core.Types$Type.go$1(Types.scala:429)
[error]     at dotty.tools.dotc.core.Types$Type.findMember(Types.scala:491)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$member$1.apply(Types.scala:407)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$member$1.apply(Types.scala:407)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.core.Types$Type.member(Types.scala:406)
[error]     at dotty.tools.dotc.core.Types$NamedType.asMemberOf(Types.scala:1355)
[error]     at dotty.tools.dotc.core.Types$NamedType.loadDenot(Types.scala:1343)
[error]     at dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:1235)
[error]     at dotty.tools.dotc.core.Types$NamedType.denotAt(Types.scala:1224)
[error]     at dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:1212)
[error]     at dotty.tools.dotc.core.Types$NamedType.symbol(Types.scala:1373)
[error]     at dotty.tools.dotc.printing.RefinedPrinter$$anonfun$toText$1.apply(RefinedPrinter.scala:125)
[error]     at dotty.tools.dotc.printing.RefinedPrinter$$anonfun$toText$1.apply(RefinedPrinter.scala:95)
[error]     at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:22)
[error]     at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:95)
[error]     at dotty.tools.dotc.printing.RefinedPrinter$$anonfun$toText$1.apply(RefinedPrinter.scala:135)
[error]     at dotty.tools.dotc.printing.RefinedPrinter$$anonfun$toText$1.apply(RefinedPrinter.scala:95)
[error]     at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:22)
[error]     at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:95)
[error]     at dotty.tools.dotc.core.Types$Type.toText(Types.scala:1025)
[error]     at dotty.tools.dotc.printing.Showable$class.show(Showable.scala:23)
[error]     at dotty.tools.dotc.core.Types$Type.show(Types.scala:73)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.signature(Denotations.scala:394)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.atSignature(Denotations.scala:421)
[error]     at dotty.tools.dotc.core.Types$TermRefWithSignature.loadDenot(Types.scala:1502)
[error]     at dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:1235)
[error]     at dotty.tools.dotc.core.Types$NamedType.denotAt(Types.scala:1224)
[error]     at dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:1212)
[error]     at dotty.tools.dotc.core.Types$NamedType.symbol(Types.scala:1373)
[error]     at dotty.tools.dotc.core.Substituters$class.subst1(Substituters.scala:31)
[error]     at dotty.tools.dotc.core.Contexts$Context.subst1(Contexts.scala:51)
[error]     at dotty.tools.dotc.core.Types$Type.subst(Types.scala:952)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.elimExistentials(Scala2Unpickler.scala:614)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.readType(Scala2Unpickler.scala:717)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$$anonfun$readTypeRef$1.apply(Scala2Unpickler.scala:795)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$$anonfun$readTypeRef$1.apply(Scala2Unpickler.scala:795)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.atReadPos(Scala2Unpickler.scala:326)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.at(Scala2Unpickler.scala:316)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.readTypeRef(Scala2Unpickler.scala:795)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.readType(Scala2Unpickler.scala:710)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler$$anonfun$9.apply(Scala2Unpickler.scala:537)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler$$anonfun$9.apply(Scala2Unpickler.scala:537)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.atReadPos(Scala2Unpickler.scala:326)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.at(Scala2Unpickler.scala:316)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler.dotty$tools$dotc$core$unpickleScala2$Scala2Unpickler$LocalUnpickler$$parseToCompletion$1(Scala2Unpickler.scala:537)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler$$anonfun$complete$1.apply$mcV$sp(Scala2Unpickler.scala:566)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler$$anonfun$complete$1.apply(Scala2Unpickler.scala:566)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler$$anonfun$complete$1.apply(Scala2Unpickler.scala:566)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.atReadPos(Scala2Unpickler.scala:326)
[error]     at dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler.complete(Scala2Unpickler.scala:566)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:161)
[error]     at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:143)
[error]     at dotty.tools.dotc.transform.TreeTransforms$AnnotationTransformer$class.transform(TreeTransform.scala:183)
[error]     at dotty.tools.dotc.transform.FirstTransform.transform(FirstTransform.scala:30)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:568)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:586)
[error]     at dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:394)
[error]     at dotty.tools.dotc.core.Symbols$Symbol.isType(Symbols.scala:410)
[error]     at dotty.tools.dotc.core.TypeErasure$$anonfun$6.apply(TypeErasure.scala:354)
[error]     at dotty.tools.dotc.core.TypeErasure$$anonfun$6.apply(TypeErasure.scala:354)
[error]     at dotty.tools.dotc.core.Scopes$Scope$$anonfun$filteredScope$1.apply(Scopes.scala:133)
[error]     at dotty.tools.dotc.core.Scopes$Scope$$anonfun$filteredScope$1.apply(Scopes.scala:132)
[error]     at scala.collection.Iterator$class.foreach(Iterator.scala:750)
[error]     at scala.collection.AbstractIterator.foreach(Iterator.scala:1202)
[error]     at dotty.tools.dotc.core.Scopes$Scope.filteredScope(Scopes.scala:132)
[error]     at dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:354)
[error]     at dotty.tools.dotc.core.TypeErasure.eraseInfo(TypeErasure.scala:385)
[error]     at dotty.tools.dotc.core.TypeErasure$.transformInfo(TypeErasure.scala:165)
[error]     at dotty.tools.dotc.transform.Erasure.transform(Erasure.scala:54)
[error]     at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:568)
[error]     at dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:1239)
[error]     at dotty.tools.dotc.core.Types$NamedType.denotAt(Types.scala:1224)
[error]     at dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:1212)
[error]     at dotty.tools.dotc.core.Types$NamedType.symbol(Types.scala:1373)
[error]     at dotty.tools.dotc.core.TypeComparer.compareNamed$1(TypeComparer.scala:156)
[error]     at dotty.tools.dotc.core.TypeComparer.dotty$tools$dotc$core$TypeComparer$$firstTry(TypeComparer.scala:187)
[error]     at dotty.tools.dotc.core.TypeComparer$$anonfun$isSubType$1.apply$mcZ$sp(TypeComparer.scala:100)
[error]     at dotty.tools.dotc.core.TypeComparer$$anonfun$isSubType$1.apply(TypeComparer.scala:92)
[error]     at dotty.tools.dotc.core.TypeComparer$$anonfun$isSubType$1.apply(TypeComparer.scala:92)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:91)
[error]     at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:85)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$$less$colon$less$1.apply$mcZ$sp(Types.scala:588)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$$less$colon$less$1.apply(Types.scala:588)
[error]     at dotty.tools.dotc.core.Types$Type$$anonfun$$less$colon$less$1.apply(Types.scala:588)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:587)
[error]     at dotty.tools.dotc.transform.Erasure$Boxing$.adaptToType(Erasure.scala:253)
[error]     at dotty.tools.dotc.transform.Erasure$Typer$$anonfun$adapt$2.apply(Erasure.scala:651)
[error]     at dotty.tools.dotc.transform.Erasure$Typer$$anonfun$adapt$2.apply(Erasure.scala:647)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:164)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.adapt(Erasure.scala:647)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1126)
[error]     at dotty.tools.dotc.transform.Erasure$Typer$$anonfun$4.apply(Erasure.scala:435)
[error]     at dotty.tools.dotc.transform.Erasure$Typer$$anonfun$4.apply(Erasure.scala:435)
[error]     at dotty.tools.dotc.core.Decorators$ListDecorator$.zipWithConserve$extension(Decorators.scala:97)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:435)
[error]     at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1039)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1079)
[error]     at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1126)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:893)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedDefDef$1.apply(Typer.scala:886)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:886)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:481)
[error]     at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1027)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1077)
[error]     at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1111)
[error]     at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1122)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:540)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:923)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:905)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:905)
[error]     at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1030)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1077)
[error]     at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1111)
[error]     at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1122)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:540)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:923)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:905)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:905)
[error]     at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1030)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1077)
[error]     at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1111)
[error]     at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1122)
[error]     at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:540)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:970)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:961)
[error]     at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
[error]     at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:961)
[error]     at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1067)
[error]     at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1079)
[error]     at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:93)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1089)
[error]     at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1087)
[error]     at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:147)
[error]     at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
[error]     at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1087)
[error]     at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1126)
[error]     at dotty.tools.dotc.transform.Erasure.run(Erasure.scala:72)
[error]     at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:268)
[error]     at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:266)
[error]     at scala.collection.immutable.List.map(List.scala:277)
[error]     at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:266)
[error]     at dotty.tools.dotc.transform.Erasure.runOn(Erasure.scala:30)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:59)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:56)
[error]     at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[error]     at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:56)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
[error]     at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
[error]     at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
[error]     at dotty.tools.dotc.Run.compileUnits(Run.scala:52)
[error]     at dotty.tools.dotc.Run.compileSources(Run.scala:49)
[error]     at dotty.tools.dotc.Run.compile(Run.scala:33)
[error]     at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
[error]     at dotty.tools.dotc.Bench$.dotty$tools$dotc$Bench$$super$doCompile(Bench.scala:38)
[error]     at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:38)
[error]     at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:36)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.immutable.Range.foreach(Range.scala:166)
[error]     at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:142)
[error]     at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
[error]     at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:136)
[error]     at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
[error]     at dotty.tools.dotc.Bench$.ntimes(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$.doCompile(Bench.scala:36)
[error]     at dotty.tools.dotc.Driver.process(Driver.scala:31)
[error]     at dotty.tools.dotc.Bench$.dotty$tools$dotc$Bench$$super$process(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$process$1.apply(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$process$1.apply(Bench.scala:53)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:30)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:142)
[error]     at scala.collection.immutable.Range.foreach(Range.scala:166)
[error]     at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:142)
[error]     at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
[error]     at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:136)
[error]     at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
[error]     at dotty.tools.dotc.Bench$.ntimes(Bench.scala:30)
[error]     at dotty.tools.dotc.Bench$.process(Bench.scala:53)
[error]     at test.CompilerTest.compileArgs(CompilerTest.scala:159)
[error]     at test.CompilerTest.compileDir(CompilerTest.scala:104)
[error]     at dotc.tests.dotty(tests.scala:139)
```
@odersky can you have a look?

**Comment by odersky** [comment]
What was the actual assertion violation?

I can't reproduce this. It's really weird. When I first pulled the branch and do sbt test I did get findMember exceptions which were NotDefinedHere violations. I made them go away by adding FutureDefsOK to the rootContext. But now I cannot even get there. I get lots of errors like this:

./src/dotty/tools/backend/jvm/CollectEntryPoints.scala:72: error: type mismatch:
 found   : sym.ThisName
 required: ?{ +: ? }
      ctx.warning(          sym.name +
                                ^
./src/dotty/tools/backend/jvm/CollectEntryPoints.scala:116: error: missing parameter type for parameter _$1, expected = ?
                if (t.resultType :: paramTypes exists (_.typeSymbol.isAbstractType))
                                                       ^
./src/dotty/tools/backend/jvm/CollectEntryPoints.scala:116: error: type mismatch:
 found   : dotty.tools.dotc.core.Symbols.Symbol
 required: ?{ isAbstractType: ? }
                if (t.resultType :: paramTypes exists (_.typeSymbol.isAbstractType))
                                                         ^
./src/dotty/tools/backend/jvm/CollectEntryPoints.scala:111: error: missing parameter type for parameter m, expected = ?
          possibles exists { m =>
                             ^
./src/dotty/tools/backend/jvm/DottyBackendInterface.scala:149: error: type mismatch:
 found   : dotty.tools.dotc.core.Symbols.Symbol
 required: ?{ requiredValue: ? }
  lazy val AnnotationRetentionSourceAttr = ctx.requiredClass("java.lang.annotation.RetentionPolicy").linkedClass.requiredValue("SOURCE")
  
followed by  a assertion failure in withType. I get the same errors with eclipse. Something seems to be wrong with the classpaths here.

**Comment by DarkDimius** [comment]
The errors that you have I haven't seen. Though travis failed with a different error.

The actual violation is 
```
[error] Test dotc.tests.dotty failed: dotty.tools.dotc.core.Denotations$NotDefinedHere: demanding denotation of <SingleDenotation of type ClassInfo(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,settings)),ScalaSettings)), class BooleanSetting)> at phase frontend(1) outside defined interval: defined periods are Period(14..18, run = 2), took 26.306 sec
```
And It I can reproduce it locally on a clean clone of the repo.
Adding  adding `FutureDefsOK` would definitely silence it, but the assertion seems strange, as the denotation seems to be not defined during typer.

**Comment by DarkDimius** [comment]
Travis reproduced the error.

**Comment by odersky** [comment]
I figured out what the previous errors are. Now we get much further, and die in BackendInterface#memberClasses.

**Comment by odersky** [comment]
@DarkDimius I suggest you merge with #544. memberClasses is causing this error, and the way it is defined here it is definitely broken.

**Comment by DarkDimius** [comment]
I believe that bb9eb23 silences actual error: we have a denotation for `BooleanSetting`, that is needed during erasure but which is not defined during typer. But this denotation should have been forced by the typer, but instead it was forced by a later phase.

**Comment by odersky** [comment]
Indeed the error was the NotDefinedHere, and the other symptom was a follow-on. I have pushed a commit that undoes the previous two commits and solves the NotDefinedHere problem. We are now 
still at the problem with memberClasses, but maybe a merge with #544 will fix that.

**Comment by DarkDimius** [comment]
@odersky please review 10610cb

File: src/dotty/tools/backend/jvm/DottyBackendInterface.scala [hunk 1]
```diff
@@ -678,8 +686,24 @@ class DottyBackendInterface()(implicit ctx: Context) extends BackendInterface{
 
     // members
     def primaryConstructor: Symbol = toDenot(sym).primaryConstructor
-    def nestedClasses: List[Symbol] = memberClasses //exitingPhase(currentRun.lambdaliftPhase)(sym.memberClasses)
-    def memberClasses: List[Symbol] = toDenot(sym).info.memberClasses.map(_.symbol).toList
+
+    /** For currently compiled classes: All locally defined classes including local classes.
```
Comment (smarter) [inline src/dotty/tools/backend/jvm/DottyBackendInterface.scala]: What's the difference between a "locally defined class" and a "local class"?

Comment (DarkDimius) [inline src/dotty/tools/backend/jvm/DottyBackendInterface.scala]: My guess is:

```
class A{
  class B
  def foo = {
    class C
  }
}
```
Class `B` is locally defined, class `C` is local.
Though the naming here is indeed confusing.

Comment (odersky) [inline src/dotty/tools/backend/jvm/DottyBackendInterface.scala]: Yes, "local classes" is Java lingo.

On Fri, May 8, 2015 at 4:43 PM, Dmitry Petrashko <notifications@github.com>
wrote:

> In src/dotty/tools/backend/jvm/DottyBackendInterface.scala
> <https://github.com/lampepfl/dotty/pull/543#discussion_r29944973>:
>
> > @@ -678,8 +686,24 @@ class DottyBackendInterface()(implicit ctx: Context) extends BackendInterface{
> >
> >      // members
> >      def primaryConstructor: Symbol = toDenot(sym).primaryConstructor
> > -    def nestedClasses: List[Symbol] = memberClasses //exitingPhase(currentRun.lambdaliftPhase)(sym.memberClasses)
> > -    def memberClasses: List[Symbol] = toDenot(sym).info.memberClasses.map(_.symbol).toList
> > +
> > +    /** For currently compiled classes: All locally defined classes including local classes.
>
> My guess is:
>
> class A{
>   class B
>   def foo = {
>     class C
>   }
> }
>
> Class B is locally defined, class C is local.
> Though the naming here is indeed confusing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/543/files#r29944973>.
>



-- 
Martin Odersky
EPFL

**Comment by odersky** [comment]
I think we can merge this now to make progress. I'll get back to the dotty deviation once this is in.

**Pull request merged**